dartvlc/internal/events.hpp

```diff
@@ -1,230 +1,223 @@
-		if (this->getDuration() > 0) {
-			this->state->isValid = this->mediaPlayer.isValid();
-			this->state->isCompleted = true;
-			this->state->position = this->getPosition();
-			this->state->duration = this->getDuration();
-			this->_onPlaylistCallback();
-			this->_completeCallback();
-		} else {
-			this->state->position = 0;
-			this->state->duration = 0;
-		}
-	}
-
-	std::function<void(float)> _volumeCallback = [=](float volume) -> void {};
-
-	std::function<void(float)> _rateCallback = [=](float rate) -> void {};
-
-	std::function<void(uint8_t* frame)> _videoCallback = [=](uint8_t* frame) -> void {};
-
-	uint8_t* _videoFrameBuffer;
-
-	void* _videoLockCallback(void** planes) {
-		planes[0] = static_cast<void*>(this->_videoFrameBuffer);
-		return nullptr;
-	}
-
-	void _videoPictureCallback(void* picture) {
-		this->_videoCallback(static_cast<uint8_t*>(this->_videoFrameBuffer));
-	}
+ public:
+  void OnOpen(std::function<void(VLC::Media)> callback) {
+    open_callback_ = callback;
+    vlc_media_player_.eventManager().onMediaChanged(
+        std::bind(&PlayerEvents::OnOpenCallback, this, std::placeholders::_1));
+  }
+
+  void OnPlay(std::function<void(void)> callback) {
+    play_callback_ = callback;
+    vlc_media_player_.eventManager().onPlaying(
+        std::bind(&PlayerEvents::OnPlayCallback, this));
+  }
+
+  void OnPause(std::function<void(void)> callback) {
+    pause_callback_ = callback;
+    vlc_media_player_.eventManager().onPaused(
+        std::bind(&PlayerEvents::OnPauseCallback, this));
+  }
+
+  void OnStop(std::function<void(void)> callback) {
+    stop_callback_ = callback;
+    vlc_media_player_.eventManager().onStopped(
+        std::bind(&PlayerEvents::OnStopCallback, this));
+  }
+
+  void OnPosition(std::function<void(int32_t)> callback) {
+    position_callback_ = callback;
+    vlc_media_player_.eventManager().onPositionChanged(std::bind(
+        &PlayerEvents::OnPositionCallback, this, std::placeholders::_1));
+  }
+
+  void OnSeekable(std::function<void(bool)> callback) {
+    seekable_callback_ = callback;
+    vlc_media_player_.eventManager().onSeekableChanged(std::bind(
+        &PlayerEvents::OnSeekableCallback, this, std::placeholders::_1));
+  }
+
+  void OnComplete(std::function<void(void)> callback) {
+    complete_callback_ = callback;
+    vlc_media_player_.eventManager().onEndReached(
+        std::bind(&PlayerEvents::OnCompleteCallback, this));
+  }
+
+  void OnVolume(std::function<void(float)> callback) {
+    volume_callback_ = callback;
+  }
+
+  void OnRate(std::function<void(float)> callback) {
+    rate_callback_ = callback;
+  }
+
+  void OnPlaylist(std::function<void(void)> callback) {
+    playlist_callback_ = callback;
+  }
+
+  void OnVideo(std::function<void(uint8_t* frame)> callback) {
+    video_callback_ = callback;
+    int32_t pitch = video_width_ * 4;
+    int32_t size = video_height_ * pitch;
+    video_frame_buffer_.reset(new uint8_t[size]);
```

Comment:
@jnschulze so, I have switched to smart pointers evevrywhere & following Google style guide for C++.

I tried to allocate buffer with `std::make_unique<uint8_t>(size)` but it wasn't working, so had to write this. I don't know if something better can be done.

Reply:
Basically, you can either write
`std::make_unique<uint8_t[]>(new uint8_t[size])`
or
`std::make_unique<uint8_t[]>(size)` which will initialize the memory